database/src/ops/block.rs

```diff
@@ -1,89 +1,321 @@
+            },
+        )
+    } else if block.block.header.major_version < 10 {
+        tables.block_info_v2s_mut().put(
+            &block.height,
+            #[allow(clippy::cast_possible_truncation)] // TODO
+            &BlockInfoV2 {
+                timestamp: block.block.header.timestamp,
+                total_generated_coins: block.generated_coins,
+                weight: block.weight as u64, // TODO
+                #[allow(clippy::cast_possible_truncation)] // TODO
+                cumulative_difficulty: block.cumulative_difficulty as u64, // TODO
+                block_hash: block.block_hash,
+                cumulative_rct_outs: todo!(), // TODO
+            },
+        )
+    } else {
+        tables.block_info_v3s_mut().put(
+            &block.height,
+            &BlockInfoV3 {
+                timestamp: block.block.header.timestamp,
+                total_generated_coins: block.generated_coins,
+                weight: block.weight as u64, // TODO
+                cumulative_difficulty: block.cumulative_difficulty,
+                block_hash: block.block_hash,
+                cumulative_rct_outs: todo!(),                    // TODO
+                long_term_weight: block.long_term_weight as u64, // TODO
+            },
+        )
+    }?;
 
-/// TODO
-pub fn block_exists() {
-    todo!()
-}
+    // Block blobs.
+    // TODO: what is a block blob in Cuprate's case?
+    tables.block_blobs_mut().put(&block.height, todo!())?;
 
-/// TODO
-pub fn get_block_hash() {
-    todo!()
-}
+    // Block heights.
+    tables
+        .block_heights_mut()
+        .put(&block.block_hash, &block.height)?;
 
-/// TODO
-pub fn get_block_height() {
-    todo!()
-}
+    // Transaction & Outputs.
+    for tx in block.txs {
+        let tx: &Transaction = &tx.tx;
 
-/// TODO
-pub fn get_block_weight() {
-    todo!()
-}
+        add_tx(
+            tx,
+            tables.block_heights_mut(),
+            tables.tx_ids_mut(),
+            tables.tx_heights_mut(),
+            tables.tx_unlock_time_mut(),
+            tables.prunable_hashes_mut(),
+            tables.prunable_tx_blobs_mut(),
+        )?;
 
-/// TODO
-pub fn get_block_already_generated_coins() {
-    todo!()
-}
+        // Output data.
+        for output in tx.prefix.outputs {
+            // Key images.
+            add_key_image(tables.key_images_mut(), output.key.as_bytes())?;
 
-/// TODO
-pub fn get_block_long_term_weight() {
-    todo!()
-}
+            // Pre-RingCT outputs.
+            if let Some(amount) = output.amount {
+                add_output(
+                    amount,
+                    &Output {
+                        key: *output.key.as_bytes(),
+                        height: todo!(),
+                        output_flags: todo!(),
+                        tx_idx: todo!(),
+                    },
+                    tables.outputs_mut(),
+                    tables.num_outputs_mut(),
+                )?;
```

Comment:
All V2 transactions produce RCT outputs not just txs which use RCT.

What this means is that V2 miner transactions are stored in the RCT table, so to decide what table to store the output in we need to look at the tx version

Reply:
> Remove the different versions of Block Info only having v3 which would just be Block Info

Will we just insert dummy values in Pre-RCT `BlockInfo`'s where needed?

Reply:
> Will we just insert dummy values in Pre-RCT BlockInfo's where needed?

no, `cumulative_rct_outs` will still be valid it will just be 0 and `long_term_weight` should be the value in `VerifiedBlockInformation`

> TxId (u64) => TxBlob (StorableVec) as the (key, value)?

yeah